src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -37,13 +37,19 @@ internal sealed class FallbackProjectManager
 
     internal void DynamicFileAdded(ProjectId projectId, ProjectKey razorProjectKey, string projectFilePath, string filePath)
     {
+        var project = _projectSnapshotManagerAccessor.Instance.GetLoadedProject(razorProjectKey);
         if (_fallbackProjectIds.Contains(projectId))
         {
-            // If this is a fallback project, then Roslyn may not track documents in the project, so these dynamic file notifications
-            // are the only way to know about files in the project.
-            AddFallbackDocument(razorProjectKey, filePath, projectFilePath);
+            // The project might have started as a fallback project, but it might have been upgraded by our getting CPS info
+            // about it. In that case, leave the CPS bits to do the work
+            if (project is ProjectSnapshot { HostProject: FallbackHostProject })
```

Comment:
I don't think this would have caused any problems, but I noticed it in the logs and it felt sketchy. Sometimes Roslyn gets called by CPS before we do, so we create fallback projects before "real" projects. This is fine, as when the real data comes in we update it correctly etc. but this fallback manager never stopped handling the projects too. At best this is double handling, adding each document twice (though the project system is smart enough to ignore the duplicate), at worse this is a future bug waiting to happen.